include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,323 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+#define BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+
+namespace boost {
+namespace urls {
+
+class decoded_view::const_iterator
+{
+    char const* begin_{nullptr};
+    char const* pos_{nullptr};
+    char const* end_{nullptr};
+    bool plus_to_space_{true};
+
+    friend decoded_view;
+
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space) noexcept
+        : begin_(str)
+        , pos_(str)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+    // end ctor
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space,
+        int) noexcept
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    const_iterator() noexcept = default;
+
+    const_iterator(const_iterator const&) noexcept = default;
+
+    const_iterator&
+    operator=(const_iterator const&) noexcept = default;
+
+    reference
+    operator*() const noexcept
```

Comment:
This should probably not be inline, and at some point we probably want to just decode it ourselves instead of calling an entire function

Reply:
P.S.: I'm literally "not sure I understand". This is not an euphemist to say I disagree.

Reply:
We want it out of line because it will be a little too big after you manually decode it. We should not use call `pct_decode_unchecked` just to decode one character because the function pays a high up-front cost just to be invoked, as it sets up things for decoding an entire buffer not one character. And it checks end_ which is not needed here. Maybe writing this implementation detail would help:

```
inline
char
pct_decode_one_unchecked(
    char const*& src,
    bool plus_to_space) noexcept;
```

and we could express the existing functions in terms of this? Anyway this is an optimization which is not important now (but do make a note of it, perhaps in an issue). What is important is that we put it out-of-line.